pkg/pool-stable/contracts/ComposableStablePoolProtocolFees.sol

```diff
@@ -235,64 +236,27 @@ abstract contract ComposableStablePoolProtocolFees is
-        // both increase.
-
-        // |-------------------------|-- postJoinExitInvariant
-        // |   increase from fees    |
-        // |-------------------------|-- original invariant * supply growth ratio (fee-less invariant)
-        // |                         |
-        // | increase from balances  |
-        // |-------------------------|-- preJoinExitInvariant
-        // |                         |
-        // |                         |  |------------------|-- postJoinExitSupply
-        // |                         |  |    BPT minted    |
-        // |                         |  |------------------|-- preJoinExitSupply
-        // |   original invariant    |  |  original supply |
-        // |_________________________|  |__________________|
-        //
-        // If the join is proportional, the invariant and supply will likewise increase proportionally,
-        // so the growth ratios (postJoinExit / preJoinExit) will be equal. In this case, we do not charge
-        // any protocol fees.
-        //
-        // If the join is non-proportional, the supply increase will be proportionally less than the invariant increase,
-        // since the BPT minted will be based on fewer tokens (because swap fees are not included). So the supply growth
-        // is due entirely to the balance changes, while the invariant growth also includes swap fees.
-        //
-        // To isolate the amount of increase by fees then, we multiply the original invariant by the supply growth
-        // ratio to get the "feeless invariant". The difference between the final invariant and this value is then
-        // the amount of the invariant due to fees, which we convert to a percentage by normalizing against the
-        // final (postJoinExit) invariant.
-        //
-        // Compute the portion of the invariant increase due to fees
-        uint256 supplyGrowthRatio = postJoinExitSupply.divDown(preJoinExitSupply);
-        uint256 feelessInvariant = preJoinExitInvariant.mulDown(supplyGrowthRatio);
+        // Compute the growth ratio between the pre- and post-join/exit balances.
+        // Note that the pre-join/exit invariant is *not* the invariant from the last join,
+        // but computed from the balances before this particular join/exit.
 
-        uint256 invariantDeltaFromFees = postJoinExitInvariant - feelessInvariant;
+        uint256 postJoinExitInvariant = StableMath._calculateInvariant(currentAmp, balances);
 
-        // To convert to a percentage of pool ownership, multiply by the rate,
-        // then normalize against the final invariant
-        uint256 protocolOwnershipPercentage = invariantDeltaFromFees.divDown(postJoinExitInvariant).mulDown(
+        // This usage of WeightedMath is intentional.
+        // `_calcDueProtocolSwapFeeBptAmount` is general to several invariants so we can use it here safely.
+        uint256 protocolFeeAmount = WeightedMath._calcDueProtocolSwapFeeBptAmount(
```

Comment:
Surely this hints at moving it out of WeightedMath?

Reply:
Yeahhhhh, I mentioned this in the OP but I'm prioritising unblocking other PRs atm. I'd to refactor a bunch of usages of this function in tests.